app/gui2/src/components/MarkdownEditor/ImagePlugin/imageNode.ts

```diff
@@ -0,0 +1,147 @@
+  key?: NodeKey | undefined
+  src: string
+}
+
+export interface UpdateImagePayload {
+  altText?: string
+}
+
+function $convertImageElement(domNode: Node): null | DOMConversionOutput {
+  if (domNode instanceof HTMLImageElement) {
+    const { alt: altText, src } = domNode
+    const node = $createImageNode({ altText, src })
+    return { node }
+  }
+  return null
+}
+
+export type SerializedImageNode = Spread<
+  {
+    altText: string
+    src: string
+  },
+  SerializedLexicalNode
+>
+
+export class ImageNode extends DecoratorNode<Component> {
+  __src: string
+  __altText: string
+
+  static getType(): string {
+    return 'image'
+  }
+
+  static clone(node: ImageNode): ImageNode {
+    return new ImageNode(node.__src, node.__altText, node.__key)
+  }
+
+  static importJSON(serializedNode: SerializedImageNode): ImageNode {
+    const { altText, src } = serializedNode
+    return $createImageNode({
+      altText,
+      src,
+    })
+  }
+
+  static importDOM(): DOMConversionMap | null {
+    return {
+      img: (_node: Node) => ({
+        conversion: $convertImageElement,
+        priority: 0,
+      }),
+    }
+  }
+
+  constructor(src: string, altText: string, key?: NodeKey) {
+    super(key)
+    this.__src = src
+    this.__altText = altText
+  }
+
+  exportDOM(): DOMExportOutput {
+    const element = document.createElement('img')
+    element.setAttribute('src', this.__src)
+    element.setAttribute('alt', this.__altText)
+    return { element }
+  }
+
+  exportJSON(): SerializedImageNode {
+    return {
+      altText: this.getAltText(),
+      src: this.getSrc(),
+      type: 'image',
+      version: 1,
+    }
+  }
+
+  getSrc(): string {
+    return this.__src
+  }
+
+  getAltText(): string {
+    return this.__altText
+  }
+
+  setAltText(altText: string): void {
+    const writable = this.getWritable()
+    writable.__altText = altText
+  }
+
+  update(payload: UpdateImagePayload): void {
+    const writable = this.getWritable()
+    const { altText } = payload
+    if (altText !== undefined) {
+      writable.__altText = altText
+    }
+  }
+
+  // View
+
+  createDOM(config: EditorConfig): HTMLElement {
+    const span = document.createElement('span')
+    const className = config.theme.image
+    if (className !== undefined) {
+      span.className = className
+    }
+    return span
+  }
+
+  updateDOM(_prevNode: ImageNode, dom: HTMLElement, config: EditorConfig): false {
+    const className = config.theme.image
+    if (className !== undefined) {
+      dom.className = className
+    }
+    return false
+  }
+
+  decorate(): Component {
+    return h(LexicalImage, {
+      src: this.__src,
+      alt: this.__altText,
+    })
+  }
+}
+
+export function $createImageNode({ altText, src, key }: ImagePayload): ImageNode {
+  return $applyNodeReplacement(new ImageNode(src, altText, key))
+}
+
+export function $isImageNode(node: LexicalNode | null | undefined): node is ImageNode {
+  return node instanceof ImageNode
+}
```

Comment:
If I understand https://github.com/facebook/lexical/discussions/1800 correctly, dollars should be for functions which mutates state (or uses $ functions) - but this does neither. Or I should read about this pattern elsewhere?

Reply:
Ok, I know now that I were wrong, the dollar is for anything _accessing_ editor state, not only modifying.

Reply:
*Reading* state is also restricted and guarded by the `$` convention. In this case it wouldn't hurt anything to call this function outside of a reading/mutating scope because it only observes an immutable quality of the value, but the convention within Lexical is to use `$` even for these predicates; I think it makes sense because identifying the node as an `ImageNode` doesn't allow you to do anything with it unless in a suitable scope.